[progress] The `total` we pass to a Progress renderable can now be `None`

DrBenton opened the pull request:
## Type of changes

- [ ] Bug fix
- [x] New feature: https://github.com/Textualize/rich/issues/1054
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [x] I've run the latest [black](https://github.com/psf/black) with default args on new code.
- [x] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [x] I've added tests for new code.
- [x] I accept that @willmcgugan may be pedantic in the code review.

## Description

With this PR we now render a "pulse" progress bar when the given `total` is None. 
Which is a new possibility we now have: this parameter used to be a `float`, and is now an `Optional[float]` - so it shouldn't bring any compatibility issues, as the type is just widened to a value that was not accepted according to the documentation, and which would raise exceptions before if it was used?

#### Design considerations

To carry the semantic of _"we don't know the 'total' value for my progress bar"_, instead of `None` we could have used other kinds of values, such as:
 - A negative value, like `total=-1`. The benefit would have been that the signature would still have been `float` , but the drawback is that the meaning of this value is less clear when debugging
 - A sentinel value, such as `total=rich.progress.UNDETERMINED` : probably the most explicit option, but a bit more cumbersome for our users than a good old `None`? :thinking:

DrBenton commented on tests/test_progress.py:
```diff
@@ -3,33 +3,32 @@
 import pytest
 
 import rich.progress
-from rich.progress_bar import ProgressBar
```
only changed these imports in order to pass the `isort` part of the pre-commit hook :slightly_smiling_face:

DrBenton commented on rich/progress.py:
```diff
@@ -40,7 +40,7 @@
 
 from . import filesize, get_console
-from .console import Console, JustifyMethod, RenderableType, Group
+from .console import Console, Group, JustifyMethod, RenderableType
```
`isort` doing its things :slightly_smiling_face:

DrBenton commented on rich/progress.py:
```diff
@@ -651,7 +651,7 @@ def __init__(
         """Gets a progress bar widget for a task."""
         return ProgressBar(
-            total=max(0, task.total),
+            total=max(0, task.total) if task.total is not None else None,
```
I had to add loads of `if task.total is not None else` here and there, now that this "total" can be `None`

willmcgugan commented:
Making total `None` may qualify as a breaking change, but the alternatives don't seem much better...

What does a progress bar with default settings display when `total` is `None` now? Can I have a screenshot please.

willmcgugan commented:
I think what we might have to do is replace that TextColumn with the percentage, with a column that displays either the percentage OR iterations per second. I'm pretty sure that is what tqdm does.

The ETA should probably not display anything if there is no total.

Basically, the user should get something that broadly does what they would expect, given the constraints. Suggest analysing what tqdm does here.

DrBenton commented:
@willmcgugan hopefully this commit addresses your comment :crossed_fingers:  :slightly_smiling_face: : [f41bda6](https://github.com/Textualize/rich/pull/2166/commits/f41bda6de0b3b62541d9c66ef3582b3d1d107c30)

The progress bar in action:
![progress-bar-with-none-total-(4)](https://user-images.githubusercontent.com/722388/163156839-17e74062-e688-4926-8096-31bf37fb5d87.gif)

willmcgugan commented on rich/progress.py:
```diff
@@ -1585,16 +1647,16 @@ def remove_task(self, task_id: TaskID) -> None:
         TimeElapsedColumn(),
         console=console,
-        transient=True,
+        transient=False,
```
I'm assuming the code below was for testing. Can you restore this so that `progress -m rich.progress` show a demonstration of multiple progress bars.

willmcgugan reviewed (changes requested):
Great! Couple of small requests.

DrBenton commented:
Wooops, sorry @willmcgugan  for the mess in the `if __name__ == "__main__":` part of the "progress.py" module - I should have reviewed myself before requesting your review again! :sweat: 

I amended the commit, hopefully addressing your 2 comments: d2830c89b7a91631a :crossed_fingers:

willmcgugan reviewed (approved):
LGTM
